test: Fix Version Message Deserialization Discrepancy

troygiorshev opened the pull request:
Inspired by theStack's in Per-Peer Message Capture [here](https://github.com/bitcoin/bitcoin/pull/19509#issuecomment-727627176).  More context [here](https://github.com/bitcoin/bitcoin/pull/19509#issuecomment-729302211) and relevant BIPs are [BIP37](https://github.com/bitcoin/bips/blob/master/bip-0037.mediawiki?rgh-link-date=2020-11-15T19%3A59%3A03Z#Extensions_to_existing_messages) and [BIP60](https://github.com/bitcoin/bips/blob/master/bip-0060.mediawiki?rgh-link-date=2020-11-15T19%3A59%3A03Z) (draft).

There is a discrepancy in the implementation of our p2p protocol between bitcoind and the testing framework.  The fRelay field is an optional field at the end of a version message as of protocol version 70001. However, when deserializing a message in bitcoind, we don't check the version to see if it can have an fRelay field or not.  Instead we unconditionally attempt to deserialize into the field.

This commit brings the testing framework in line with the implementation in core.

This matters for a version message with the following fields:

Version = 60000
fRelay = 1

Bitcoind would deserialize this into a version message with Version=60000 and fRelay=1, whereas (before this PR) our testing framework would deserialize this into a version message with Version=60000 and fRelay=0.

~~Concept ACK needed.  Alternative: Change behavior in core?~~  Got a few 😄

troygiorshev committed:
Unconditionally check for fRelay field in test framework

There is a discrepancy in the implementation of our p2p protocol between
bitcoind and the testing framework.  The fRelay field is an optional
field at the end of a version message as of protocol version 70001.
However, when deserializing a message in bitcoind, we don't check the
version to see if it should have an fRelay field or not.  Instead we
unconditionally attempt to deserialize into the field.

This commit brings the testing framework in line with the implementation
in core.

This matters for a version message with the following fields:

Version = 60000
fRelay = 1

Bitcoind would deserialize this into a version message with
Version=60000 and fRelay=1, whereas (before this commit) our testing
framework would deserialize this into a version message with
Version=60000 and fRelay=0.

sipa commented:
Concept ACK

MarcoFalke commented on test/functional/test_framework/messages.py:
```diff
@@ -1050,13 +1050,11 @@ def deserialize(self, f):
-        else:
+        # Relay field is optional for version 70001 onwards
+        # But, unconditionally check it to match behaviour in bitcoind
+        try:
```
```suggestion
```

Looks like you still treat it as optional, no?

MarcoFalke replied:
On a second thought,  I don't think this generally works the way you want it to work. (See my latest comment on your other pull request: https://github.com/bitcoin/bitcoin/pull/19509#issuecomment-729462566 )

So I've unresolved this conversation again.

MarcoFalke commented:
review ACK 0a7eedc35afd5c73d5ab7d2b5b305f8c795e5b42

theStack reviewed (approved):
Code Review ACK 0a7eedc35afd5c73d5ab7d2b5b305f8c795e5b42
Thanks for fixing this!

laanwj commented on test/functional/test_framework/messages.py:
```diff
@@ -1050,13 +1050,11 @@ def deserialize(self, f):
+        # But, unconditionally check it to match behaviour in bitcoind
+        try:
+            self.nRelay = struct.unpack("<b", f.read(1))[0]
+        except:
```
Please don't use a blanket except here, but catch the specific exception `struct.error`. This could potentially mask another error, even "KeyboardInterrupt" (ctrl-C).

MarcoFalke replied:
Would probably be best to remove the `except` wholesale.  https://github.com/bitcoin/bitcoin/pull/20411/files#r528499393

jnewbery replied:
This isn't new. It's simply moving the try-except out of an if block, but there's no harm in fixing it up in this PR.

troygiorshev commented:
`git range-diff master 0a7eedc 68500bb`

Trivial rebase since 19509 needs a rebase and will be based off of this PR.

jnewbery reviewed (commented):
ACK 68500bb48b63841591fae475e0b4760c021a6af6

This is actually more spec compliant. BIP 37, which introduced the fRelay field to the version message, does not specify a minimum p2p version: https://github.com/bitcoin/bips/blob/master/bip-0037.mediawiki.

I'm happy to reACK a version of this that narrows the exception catching to just struct.error exceptions.

jnewbery committed:
[test] Move MY_RELAY to p2p.py

messages.py is for message and primitive data structures. Specifics
about the test framework's p2p implementation should be in p2p.py.

Also rename to P2P_VERSION_RELAY. Also rename msg_version.nRelay to
relay. In Bitcoin Core, this is referred to as fRelay, since it's a
bool, so this field has always been misnamed.

jnewbery committed:
[test] Move MIN_VERSION_SUPPORTED to p2p.py

The messages.py module should contain code and helpers for
[de]serializing p2p messages. Specific usage of those messages should
be in p2p.py. Therefore move MIN_VERSION_SUPPORTED to p2p.py.

Also rename to MIN_P2P_VERSION_SUPPORTED to distinguish it from
other versioning used in Bitcoin/Bitcoin Core.

jnewbery committed:
[test] Move MY_VERSION to p2p.py

The messages.py module should contain code and helpers for
[de]serializing p2p messages. Specific usage of those messages should
be in p2p.py. Therefore move MY_VERSION to p2p.py.

Also rename to P2P_VERSION to distinguish it from
other versioning used in Bitcoin/Bitcoin Core.

Also always set the nVersion field in CBlockLocator to 0 and ignore the
field in deserialized messages. The field is not currently used for
anything in Bitcoin Core.

jnewbery committed:
[test] Move MY_SUBVERSION to p2p.py

The messages.py module should contain code and helpers for
[de]serializing p2p messages. Specific usage of those messages should
be in p2p.py. Therefore move MY_SUBVERSION to p2p.py.

Also rename to P2P_SUBVERSION.

jnewbery committed:
[test] Add P2P_SERVICES to p2p.py

The messages.py module should contain code and helpers for
[de]serializing p2p messages. Specific usage of those messages should
be in p2p.py. Therefore specify the nServices value in the calling code,
not in the messages.py module.

jnewbery committed:
[test] Change P2P_SUBVERSION to be a string

This forces the user to convert to bytes to encode it in a version
message, which would result in an easy-to-detect type error if omitted.

Also add a check that new connections from the test framework to the
node have the correct user agent string. Again, this makes bugs easier
to detect if the user agent string ever changes.

troygiorshev commented:
Fixed the error handling here, sorta.  The error wouldn't be in the struct unpacking as suggested, it would be in the file reading.  (The try/catch is expecting the last byte to only optionally exist)  However, it's not clear to me that `f.read()` fails with the same error on all systems.  It's a `ValueError` in at least some systems, but I'm worried about breaking things, especially with this not being the main point of this PR.

I'm about to rebase this onto 20524 to prevent a merge conflict.

I realize now that this doesn't completely fix the problems in 19509 that I was originally looking at, but this is still necessary.  I'll make the remaining changes in 19509, so after rebasing this will be ready for review.

troygiorshev commented:
Rebased onto #20524.

This is ready for re-review.

jnewbery commented:
@troygiorshev are you planning to rebase this, or should we mark it as up for grabs?

jnewbery commented:
Marking this as up for grabs

jnewbery commented:
The first 6 commits in this PR were merged in #20524, so it's only the final _Unconditionally check for fRelay field in test framework_ commit that would need to be included in a new PR.
